### PR TITLE
 Add ViewGroup HierarchyChange bindings

### LINKED ISSUE #79: RxViewGroup hierarchy change binding

### PR DESCRIPTION
**Pull request by xfumihiro**
Closes #79.

### CODE REVIEW

**Comment by JakeWharton** [comment]
Thanks. Will look tomorrow.

**Comment by xfumihiro** [comment]
:+1:

**Comment by JakeWharton** [comment]
(Kicking for CI)

File: rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChildViewRemoveEvent.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import android.support.annotation.CheckResult;
+import android.support.annotation.NonNull;
+import android.view.View;
+import android.view.ViewGroup;
+
+/**
+ * A child view remove event on a viewgroup.
+ */
+public class ViewGroupHierarchyChildViewRemoveEvent extends ViewGroupHierarchyChangeEvent {
```
Comment (JakeWharton) [inline rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChildViewRemoveEvent.java]: final

File: rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChildViewAddEvent.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import android.support.annotation.CheckResult;
+import android.support.annotation.NonNull;
+import android.view.View;
+import android.view.ViewGroup;
+
+/**
+ * A child view add event on a viewgroup.
+ */
+
+public class ViewGroupHierarchyChildViewAddEvent extends ViewGroupHierarchyChangeEvent {
```
Comment (JakeWharton) [inline rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChildViewAddEvent.java]: final

File: rxbinding/src/androidTest/java/com/jakewharton/rxbinding/view/RxViewGroupTestActivity.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  ViewGroup viewGroup;
+  View child;
+
+  @Override protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    viewGroup = new LinearLayout(this);
+    child = new View(this);
+    setContentView(viewGroup);
+  }
+}
```
Comment (JakeWharton) [inline rxbinding/src/androidTest/java/com/jakewharton/rxbinding/view/RxViewGroupTestActivity.java]: nit: missing trailing newline

File: rxbinding/src/androidTest/java/com/jakewharton/rxbinding/view/RxViewGroupTestActivity.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.jakewharton.rxbinding.view;
+
+import android.app.Activity;
+import android.os.Bundle;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.LinearLayout;
+
+public final class RxViewGroupTestActivity extends Activity {
+  ViewGroup viewGroup;
```
Comment (JakeWharton) [inline rxbinding/src/androidTest/java/com/jakewharton/rxbinding/view/RxViewGroupTestActivity.java]: Just make this type `LinearLayout` so you don't have to cast in the test `before()` method

File: rxbinding/src/androidTest/java/com/jakewharton/rxbinding/view/RxViewGroupTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  @Test @UiThreadTest public void childViewEvents() {
+    RecordingObserver<ViewGroupHierarchyChangeEvent> o = new RecordingObserver<>();
+    Subscription subscription = RxViewGroup.changeEvents(viewGroup)
+        .subscribeOn(AndroidSchedulers.mainThread())
+        .subscribe(o);
+    o.assertNoMoreEvents(); // No initial value.
+
+    instrumentation.runOnMainSync(new Runnable() {
```
Comment (JakeWharton) [inline rxbinding/src/androidTest/java/com/jakewharton/rxbinding/view/RxViewGroupTest.java]: For `@UiThreadTest` you don't need `runOnMainSync`

File: rxbinding/src/androidTest/java/com/jakewharton/rxbinding/view/RxViewGroupTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  @Before public void setUp() {
+    RxViewGroupTestActivity activity = activityRule.getActivity();
+    viewGroup = (LinearLayout) activity.viewGroup;
+    child = activity.child;
+  }
+
+  @Test @UiThreadTest public void childViewEvents() {
+    RecordingObserver<ViewGroupHierarchyChangeEvent> o = new RecordingObserver<>();
+    Subscription subscription = RxViewGroup.changeEvents(viewGroup)
+        .subscribeOn(AndroidSchedulers.mainThread())
```
Comment (JakeWharton) [inline rxbinding/src/androidTest/java/com/jakewharton/rxbinding/view/RxViewGroupTest.java]: You don't need `subscribeOn` for `@UiThreadTest`

File: rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChangeEventOnSubscribe.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import com.jakewharton.rxbinding.internal.MainThreadSubscription;
+import rx.Observable;
+import rx.Subscriber;
+
+import static com.jakewharton.rxbinding.internal.Preconditions.checkUiThread;
+
+final class ViewGroupHierarchyChangeEventOnSubscribe implements Observable.OnSubscribe<ViewGroupHierarchyChangeEvent> {
+  private final ViewGroup viewGroup;
+
+  public ViewGroupHierarchyChangeEventOnSubscribe(ViewGroup viewGroup) {
```
Comment (JakeWharton) [inline rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChangeEventOnSubscribe.java]: nit: doesn't need to be public

**Comment by JakeWharton** [comment]
Mostly looks good. Just some small nits.

File: rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChildViewAddEvent.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+import android.support.annotation.CheckResult;
+import android.support.annotation.NonNull;
+import android.view.View;
+import android.view.ViewGroup;
+
+/**
+ * A child view add event on a viewgroup.
+ */
+
```
Comment (JakeWharton) [inline rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChildViewAddEvent.java]: nit: remove empty line

File: rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChildViewAddEvent.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
+package com.jakewharton.rxbinding.view;
+
+import android.support.annotation.CheckResult;
+import android.support.annotation.NonNull;
+import android.view.View;
+import android.view.ViewGroup;
+
+/**
+ * A child view add event on a viewgroup.
```
Comment (JakeWharton) [inline rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChildViewAddEvent.java]: This Javadoc does not match the other events which include a warning. Also "viewgroup" is not a word and should probably be a `{@link ..}`

File: rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChildViewRemoveEvent.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
+package com.jakewharton.rxbinding.view;
+
+import android.support.annotation.CheckResult;
+import android.support.annotation.NonNull;
+import android.view.View;
+import android.view.ViewGroup;
+
+/**
+ * A child view remove event on a viewgroup.
```
Comment (JakeWharton) [inline rxbinding/src/main/java/com/jakewharton/rxbinding/view/ViewGroupHierarchyChildViewRemoveEvent.java]: Same Javadoc concern

File: rxbinding/src/main/java/com/jakewharton/rxbinding/view/RxViewGroup.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package com.jakewharton.rxbinding.view;
+
+import android.support.annotation.CheckResult;
+import android.support.annotation.NonNull;
+import android.view.ViewGroup;
+import rx.Observable;
+
+public final class RxViewGroup {
+  /**
+   * Create an observable of hierarchy change events for {@code viewGroup}.
```
Comment (JakeWharton) [inline rxbinding/src/main/java/com/jakewharton/rxbinding/view/RxViewGroup.java]: This Javadoc does not match the other operators which include a warning and details of the use of a setter.

File: rxbinding/src/main/java/com/jakewharton/rxbinding/view/RxViewGroup.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
+package com.jakewharton.rxbinding.view;
+
+import android.support.annotation.CheckResult;
+import android.support.annotation.NonNull;
+import android.view.ViewGroup;
+import rx.Observable;
+
+public final class RxViewGroup {
```
Comment (JakeWharton) [inline rxbinding/src/main/java/com/jakewharton/rxbinding/view/RxViewGroup.java]: Missing type-level Javadoc to match the other `RxFoo` types

**Comment by JakeWharton** [comment]
Looks like this conflicts with the changes on `master`. Can you rebase?

**Comment by JakeWharton** [comment]
Actually I'll just do it while merging.

**Comment by JakeWharton** [comment]
Merged manually.

**Comment by xfumihiro** [comment]
thanks! :+1:

**Comment by xfumihiro** [comment]
@JakeWharton you have remove the RxViewGroupTestActivity which makes the test failed.

**Comment by xfumihiro** [comment]
I've rebased the branch to ae288382702c5af2e1ad389899bab906ef2e4236.
Should I send another PR?

**Comment by JakeWharton** [comment]
It wasn't needed. I modified the test to work without it.

**Comment by xfumihiro** [comment]
Okay, but I got IllegalStateException now.

java.lang.IllegalStateException: Expected no more events but got OnError[java.lang.IllegalStateException: Must be called from the main thread. Was: Thread[Instr: com.jakewharton.rxbinding.RxBindingTestRunner,5,main]]
at com.jakewharton.rxbinding.RecordingObserver.assertNoMoreEvents(RecordingObserver.java:63)
at com.jakewharton.rxbinding.view.RxViewGroupTest.childViewEvents(RxViewGroupTest.java:29)

**Comment by JakeWharton** [comment]
#122

**Comment by xfumihiro** [comment]
:+1:

**Pull request closed without merging**